### PR TITLE
feat(inputs.systemd_units): add systemctl show subcommand to allow gathering more information about units

### LINKED ISSUE #14521: Extend systemd_unit plugin to gather restart-count and other information
### Use Case

The `systemd_units` plugin is missing some information, that is essential for my use case:

- The restart count of the unit
- If the unit is enabled or not
- The preset-state of the unit

### Expected behavior

It would be great if more information about a systemd unit would be available. I'd suggest "restart count", "memory usage", enable-, preset-state and the last exit status of the main process.

### Actual behavior

I had a look at the implementation of the `systemd_units` plugin and noticed that it calls `systemctl list-units`. Using this interface has two drawbacks in my opinion:
- Its output is hard to parse because the output is meant for human consumption.
- Some vital information (like the restart count) is missing.

### Additional info

I'm far from fluent in Go, but I had a shot converting the `systemd_units` plugin from using `systemctl list-units` to `systemctl show`. This call is able to return much more information, supports the same filtering and returns the result as easy to parse name-value-pairs.
This works quite well. Before creating a PR to change the `systemd_units` plugin, I need some advice:
- Would the conversion from `systemctl list-units` to `systemctl show` be appreciated? It would allow gathering much more metrics about the systemd units.
- How should I implement new measurements like the restart count? Should I use a feature-flag in the config so that users upgrading experience no change in functionality, or should new values just be added?

Your advice is much appreciated.

### PR DESCRIPTION
**Pull request by FlashSystems**
## Summary

The systemd_units plugin is missing some information, that is essential for my use case:
- The restart count of the unit
- If the unit is enabled or not
- The preset-state of the unit

## Checklist

- [x] No AI generated code was used in this PR

## Related issues

resolves #14521

## Details

I had a look at the implementation of the systemd_units plugin and noticed that it calls `systemctl list-units`. Using this interface has two drawbacks in my opinion:
- Its output is hard to parse because the output is meant for human consumption.
- Some vital information (like the restart count) is missing.

Like discussed in issue #14521 this PR keeps the existing functionality the same and sets it as the default. There should be no observable change in behavior with this PR.

The functionality and tests are split into two files prefixed with `subcommand_list` for the old behavior and `subcommand_show` for the new. Some code that is shared between the two subcommand implementations was put into files prefixed with `shared_`.

The user can enable the usage of the `systemctl show` command by setting the `subcommand` configuration option as outlined in the `README.md` file.

I'm far from fluent in Go, but tried my best to make the code clean, DRY and maintainable. Your suggestions on how to improve it are highly welcome.

### CODE REVIEW

**Comment by FlashSystems** [comment]
Updated: Rebase on current master.

**Comment by FlashSystems** [comment]
Updated: Fix new linter error `appendCombine`

File: plugins/inputs/systemd_units/README.md [hunk 1]
```diff
@@ -29,22 +57,30 @@ See the [CONFIGURATION.md][CONFIGURATION.md] for more details.
...
-  #
+
   ## Filter for a specific pattern, default is "" (i.e. all), other possible
   ## values are valid pattern for systemctl, e.g. "a*" for all units with
   ## names starting with "a"
   # pattern = ""
-  ## pattern = "telegraf* influxdb*"
-  ## pattern = "a*"
+  # pattern = "telegraf* influxdb*"
+  # pattern = "a*"
```
Comment (powersj) [inline plugins/inputs/systemd_units/README.md:77]: In general the config options should show the default value. We like to avoid adding examples below, and rather keep them in the comments. Let's remove these and either keep the current comments or add additional examples.

File: plugins/inputs/systemd_units/sample.conf [hunk 2]
```diff
@@ -2,15 +2,21 @@
...
-  #
+
   ## Filter for a specific pattern, default is "" (i.e. all), other possible
   ## values are valid pattern for systemctl, e.g. "a*" for all units with
   ## names starting with "a"
   # pattern = ""
-  ## pattern = "telegraf* influxdb*"
-  ## pattern = "a*"
+  # pattern = "telegraf* influxdb*"
+  # pattern = "a*"
```
Comment (powersj) [inline plugins/inputs/systemd_units/sample.conf:22]: As above, let's remove these examples.

File: plugins/inputs/systemd_units/shared_test.go [hunk 3]
```diff
@@ -0,0 +1,82 @@
+package systemd_units
```
Comment (powersj) [inline plugins/inputs/systemd_units/shared_test.go:1]: Thanks for some additional tests, however let's keep the general tests in `systemd_units_test.go` please and remove this file.

File: plugins/inputs/systemd_units/shared_maps.go [hunk 4]
```diff
@@ -0,0 +1,98 @@
+package systemd_units
```
Comment (powersj) [inline plugins/inputs/systemd_units/shared_maps.go:1]: These could stay in systemd_units.go, no?

**Review by powersj — COMMENTED** [review]
Thank you for working through this and putting up a PR! Few comments, but nothing major.

**Review by powersj — APPROVED** [review]
This is pretty awesome! Thank you

File: plugins/inputs/systemd_units/subcommand_list.go [hunk 5]
```diff
@@ -0,0 +1,88 @@
...
+	"bufio"
+	"bytes"
+	"fmt"
+	"strings"
+
+	"github.com/influxdata/telegraf"
+)
+
+// Gather parses systemctl outputs and adds counters to the Accumulator
+func parseListUnits(_ *SystemdUnits, buffer *bytes.Buffer, acc telegraf.Accumulator) error {
```
Comment (srebhan) [inline plugins/inputs/systemd_units/subcommand_list.go:14]: Please make the accumulator the first parameter. This is an unwritten rule (context, accumulator, remaining) and allows to spot the usage of the accumulator easily.

File: plugins/inputs/systemd_units/subcommand_list.go [hunk 6]
```diff
@@ -0,0 +1,88 @@
...
+	scanner := bufio.NewScanner(buffer)
+	for scanner.Scan() {
+		line := scanner.Text()
+
+		data := strings.Fields(line)
+		if len(data) < 4 {
+			acc.AddError(fmt.Errorf("Error parsing line (expected at least 4 fields): %s", line))
```
Comment (srebhan) [inline plugins/inputs/systemd_units/subcommand_list.go:21]: Error messages should start lowercase so how about
```suggestion
			acc.AddError(fmt.Errorf("parsing line failed (expected at least 4 fields): %s", line))
```
Same for all other error messages here.

File: plugins/inputs/systemd_units/subcommand_list.go (hunk 5 above)
Comment (srebhan) [inline plugins/inputs/systemd_units/subcommand_list.go:14]: This function never returns an error, does it? If not, you might want to remove the return value...

File: plugins/inputs/systemd_units/subcommand_show.go [hunk 7]
```diff
@@ -0,0 +1,149 @@
...
+	"MemoryCurrent":     {valueName: "mem_current", valueMap: nil},
+	"MemoryPeak":        {valueName: "mem_peak", valueMap: nil},
+	"MemorySwapCurrent": {valueName: "swap_current", valueMap: nil},
+	"MemorySwapPeak":    {valueName: "swap_peak", valueMap: nil},
+	"MemoryAvailable":   {valueName: "mem_avail", valueMap: nil},
+	"MainPID":           {valueName: "pid", valueMap: nil},
+}
+
+// Gather parses systemctl outputs and adds counters to the Accumulator
+func parseShow(_ *SystemdUnits, buffer *bytes.Buffer, acc telegraf.Accumulator) error {
```
Comment (srebhan) [inline plugins/inputs/systemd_units/subcommand_show.go:48]: Neither this function nor the `parseListUnits` function uses the `SystemdUnits` parameter nor does any of those return an error, so how about the following
```suggestion
func parseShow(acc telegraf.Accumulator, buffer *bytes.Buffer) {
```

File: plugins/inputs/systemd_units/systemd_units.go [hunk 8]
```diff
@@ -128,109 +119,80 @@ var subMap = map[string]int{
 	"elapsed": 0x00a0,
 }
 
+// SystemdUnits is a telegraf plugin to gather systemd unit status
+type SystemdUnits struct {
+	Timeout       config.Duration
```
Comment (srebhan) [inline plugins/inputs/systemd_units/systemd_units.go:124]: Please also add the toml tag here
```suggestion
	Timeout       config.Duration `toml:"timeout"`
```

File: plugins/inputs/systemd_units/systemd_units.go [hunk 9]
```diff
@@ -128,109 +119,80 @@ var subMap = map[string]int{
...
-		}
-
-		acc.AddFields(measurement, fields, tags)
+	switch s.SubCommand {
+	case "show":
+		subCommandInfo = initSubcommandShow()
+	case "list-units":
+		subCommandInfo = initSubcommandListUnits()
+	default:
+		return fmt.Errorf("invalid value for 'subcommand'")
```
Comment (srebhan) [inline plugins/inputs/systemd_units/systemd_units.go:160]:
```suggestion
		return fmt.Errorf("invalid value for 'subcommand': %s", s.SubCommand)
```

File: plugins/inputs/systemd_units/systemd_units.go [hunk 10]
```diff
@@ -128,109 +119,80 @@ var subMap = map[string]int{
...
 	}
-	params = append(params,
-		"--all", "--plain",
-		// add type as configured in config
-		"--type="+unitType,
-		"--no-legend",
-	)
-	cmd := exec.Command(systemctlPath, params...)
+
+	cmd := exec.Command(systemctlPath, *s.commandParams...)
```
Comment (srebhan) [inline plugins/inputs/systemd_units/systemd_units.go:178]: How about building the command at `Init` time?

**Review by srebhan — COMMENTED** [review]
Thanks for your contribution @FlashSystems! Just a few smaller comments from my side...

File: plugins/inputs/systemd_units/subcommand_show.go (hunk 7 above)
Comment (FlashSystems) [reply plugins/inputs/systemd_units/subcommand_show.go:48]: The unused `SystemdUnits` structure and return value are there to ease extending the parser functions. I thought that later on it may be necessary to access configuration options or return an error when parsing.

If this currently unused noise is a non-idiomatic, I'll remove the parameter and return value. Just wanted to make sure you understand my thoughts behind the decision to put them in.

Please give me a short feedback if these should be removed or kept.

Comment (srebhan) [reply plugins/inputs/systemd_units/subcommand_show.go:48]: I understand. However, we can easily add those back later if need be. We already have way-too-many "this is for a future that never happens" things in the code-base. ;-)

File: plugins/inputs/systemd_units/systemd_units.go (hunk 10 above)
Comment (FlashSystems) [reply plugins/inputs/systemd_units/systemd_units.go:178]: The [go documentation](https://pkg.go.dev/os/exec#Cmd) states: "A Cmd cannot be reused after calling its Run, Output or CombinedOutput methods."
I thought it would not be possible to reuse the `Cmd` instance returned by `exec.Command`. Did I miss something?

**Comment by FlashSystems** [comment]
Updated the PR to resolved review comments.

File: plugins/inputs/systemd_units/systemd_units.go (hunk 10 above)
Comment (srebhan) [reply plugins/inputs/systemd_units/systemd_units.go:178]: You are absolutely right. Sorry for the noise!

**Review by srebhan — APPROVED** [review]
Thank you for the awesome work @FlashSystems!

**Comment by srebhan** [comment]
@powersj as there where quite some changes here, do you want to take a second look? If not, feel free to merge right away...

**Review by powersj — APPROVED** [review]
Thank you again!
